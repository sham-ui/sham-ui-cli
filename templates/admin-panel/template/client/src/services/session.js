import { DI, inject } from 'sham-ui';

export default class Session {
    @inject router;
    @inject store;
    @inject( 'session:storage' ) data;

    constructor() {
        DI.bind( 'session', this );
    }

    login( email, password ) {
        return this.store.login( { email, password } ).then( ( { Email, Name } ) => {
            this.data.email = Email;
            this.data.name = Name;
        } ).then(
            ::this.resetSessionValidation
        ).then(
            () => this.validateSessionPromise
        );
    }

    get validateSessionPromise() {
        if ( undefined === this._validateSessionPromise ) {
            this._validateSessionPromise = this.validateSession();
        }
        return this._validateSessionPromise;
    }

    resetSessionValidation() {
        this._validateSessionPromise = undefined;
    }

    validateSession() {
        this.data.sessionValidated = false;
        return this.store.validSession().then(
            ( { Email, Name } ) => {
                this.data.sessionValidated = true;
                this.data.isAuthenticated = true;
                this.data.email = Email;
                this.data.name = Name;

                // Manual run sync for guaranteed update Layout
                // component before promise resolved
                this.data.sync();
                return true;
            },
            () => {
                this.data.sessionValidated = true;
                this.data.isAuthenticated = false;
                this._goToLoginPage();
                return false;
            }
        );
    }

    logout() {
        this.store.logout().then(
            () => {
                this.resetSessionValidation();
                this._goToLoginPage();
            }
        );
    }

    _isRestrictedPage() {
        const page = this.router.lastRouteResolved();
        return undefined !== page && ![
            'signup',
            'login'
        ].includes( page.name );
    }

    _goToLoginPage() {
        this.data.isAuthenticated = false;
        this.data.email = '';
        this.data.name = '';
        if ( this._isRestrictedPage() ) {
            requestAnimationFrame(
                () => this.router.navigate(
                    this.router.generate( 'login', {} )
                )
            );
        }
    }
}
