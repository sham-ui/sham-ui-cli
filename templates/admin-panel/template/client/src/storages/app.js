import createStorage from 'sham-ui-data-storage';

export const { useStorage } = createStorage( {
    routerResolved: false,
    tokenLoaded: false
}, {
    DI: 'app:storage',
    sync: true
} );
