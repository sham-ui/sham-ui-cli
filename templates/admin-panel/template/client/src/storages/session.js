import createStorage from 'sham-ui-data-storage';

export const { useStorage } = createStorage( {
    name: '',
    email: '',
    sessionValidated: false,
    isAuthenticated: false,
    isSuperuser: false
}, {
    DI: 'session:storage'
} );
