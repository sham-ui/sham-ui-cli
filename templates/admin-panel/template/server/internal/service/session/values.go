package session

import (
	"github.com/gorilla/sessions"
)

type valueNotFoundError struct {
	key string
}

func (e valueNotFoundError) Error() string {
	return "value not found: " + e.key
}

func (e valueNotFoundError) Is(target error) bool {
	//nolint:errorlint
	_, ok := target.(valueNotFoundError)
	return ok
}

type valueTypeError struct {
	key string
}

func (e valueTypeError) Error() string {
	return "value type error: " + e.key
}

func (e valueTypeError) Is(target error) bool {
	//nolint:errorlint
	_, ok := target.(valueTypeError)
	return ok
}

func newValueNotFoundError(key string) error {
	return valueNotFoundError{key: key}
}

func newValueTypeError(key string) error {
	return valueTypeError{key: key}
}

type sessionValue interface {
	~int | ~string | ~bool
}

func getValue[T sessionValue](s *sessions.Session, key string) (T, error) { //nolint:ireturn
	var zeroValue T
	value, ok := s.Values[key]
	if !ok {
		return zeroValue, newValueNotFoundError(key)
	}
	typedValue, ok := value.(T)
	if !ok {
		return zeroValue, newValueTypeError(key)
	}
	return typedValue, nil
}
