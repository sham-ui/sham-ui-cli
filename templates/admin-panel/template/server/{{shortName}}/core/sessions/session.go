package sessions

import "github.com/gorilla/sessions"

type Session struct {
	ID          string
	Email       string
	Name        string
	IsSuperuser bool

	rawSession *sessions.Session
}

func (s *Session) GetRawSession() *sessions.Session {
	return s.rawSession
}
