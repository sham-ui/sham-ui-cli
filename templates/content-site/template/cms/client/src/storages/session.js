import createStorage from 'sham-ui-data-storage';

export const { storage, useStorage } = createStorage( {
    name: '',
    email: '',
    sessionValidated: false,
    isAuthenticated: false,
    isSuperuser: false
}, {
    DI: 'session:storage'
} );
