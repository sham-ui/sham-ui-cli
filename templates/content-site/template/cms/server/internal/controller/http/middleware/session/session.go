package session

import (
	"cms/internal/controller/http/request"
	"cms/internal/model"
	"cms/pkg/logger"
	"net/http"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

type Service interface {
	Get(r *http.Request) (*model.Session, error)
}

type session struct {
	srv Service
}

func (sess *session) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(
		func(rw http.ResponseWriter, r *http.Request) {
			memberSess, err := sess.srv.Get(r)
			if err == nil {
				ctx := r.Context()

				// Save session to trace context
				span := trace.SpanFromContext(ctx)
				if span.SpanContext().IsValid() {
					span.SetAttributes(
						attribute.String("member_id", string(memberSess.MemberID)),
					)
				}

				// Save session to logger context
				log := logger.Load(ctx).WithValues(
					"member_id", string(memberSess.MemberID),
				)
				ctx = logger.Save(r.Context(), log)

				// Save session to request context
				ctx = request.SaveSessionToContext(ctx, memberSess)
				r = r.WithContext(ctx)
			}
			next.ServeHTTP(rw, r)
		},
	)
}

func New(srv Service) *session {
	return &session{
		srv: srv,
	}
}
