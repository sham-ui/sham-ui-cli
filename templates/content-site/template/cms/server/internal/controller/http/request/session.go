package request

import (
	"cms/internal/model"
	"context"
)

type contextKeySession struct{}

func SessionFromContext(ctx context.Context) (*model.Session, bool) {
	sess, ok := ctx.Value(contextKeySession{}).(*model.Session)
	return sess, ok
}

func SaveSessionToContext(ctx context.Context, session *model.Session) context.Context {
	return context.WithValue(ctx, contextKeySession{}, session)
}
