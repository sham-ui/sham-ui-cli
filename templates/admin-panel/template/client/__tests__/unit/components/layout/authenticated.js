import { createDI } from 'sham-ui';
import * as directives from 'sham-ui-directives';
import hrefto from 'sham-ui-router/lib/href-to';
import LayoutAuthenticated  from '../../../../src/components/layout/authenticated.sfc';
import renderer, { compile } from 'sham-ui-test-helpers';
import { storage } from '../../../../src/storages/session';


it( 'renders correctly', () => {
    const DI = createDI();

    storage( DI ).name = 'Test member';

    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' ),
        activePageComponent: compile``,
        storage: {
            url: '/',
            addWatcher() {}
        }
    } );

    const meta = renderer( LayoutAuthenticated, {}, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
