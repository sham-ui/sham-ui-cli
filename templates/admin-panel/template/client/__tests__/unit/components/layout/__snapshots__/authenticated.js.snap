// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "Constructor": "LayoutAuthenticated",
  "Options": Object {
    "ID": "component",
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "menuCollapsed": false,
    "routerData": Object {
      "activePageComponent": null,
      "activePageOptions": null,
      "name": "",
      "pageLoaded": true,
      "params": Object {},
      "url": "",
    },
    "smallScreenSize": false,
    "wrapperClasses": "",
  },
  "html": "
<div class=\\"\\">
  <div class=\\"logo\\">
    <a href=\\"/\\"><span>[</span> {{ logoText }} <span>]</span></a>
  </div>
  <div class=\\"sideleft\\">
    <label class=\\"sidebar-label\\">Navigation</label>
    <div class=\\"menu\\">
      <a href=\\"/\\" class=\\"active\\">
        <div class=\\"menu-item\\">
          <i class=\\"icon-home-outline menu-item-icon\\"></i><span class=\\"menu-item-label\\">Dashboard</span>
        </div>
      </a>
      <!--if-->
    </div>
  </div>
  <div class=\\"header\\">
    <div class=\\"header-left\\">
      <div class=\\"navicon-left\\">
        <a>
          <i class=\\"icon-menu\\"></i>
        </a>
      </div>
    </div>
    <div class=\\"header-right\\">
      <div class=\\"dropdown\\">
        <a class=\\"link-profile\\"><span class=\\"logged-name\\">Test member</span></a>
        <div class=\\"dropdown-menu \\">
          <ul class=\\"user-profile-nav\\">
            <li>
              <a href=\\"/\\">
                <i class=\\"icon-cog\\"></i> Settings
              </a>
            </li>
            <li>
              <a href=\\"#\\">
                <i class=\\"icon-logout\\"></i> Sign Out
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div class=\\"mainpanel\\">
    <!--ActivePageContainer-->
  </div>
</div>
",
}
`;
